src/app/clusters/occupancy-sensor-server/occupancy-sensor-server.cpp

```diff
@@ -16,14 +16,144 @@
+    }
+    return holdTimeLimitsStruct;
+}
+
+CHIP_ERROR
+HoldTimeLimitsManager::HoldTimeLimits::SetHoldTimeLimitsStruct(EndpointId endpoint,
+                                                               Structs::HoldTimeLimitsStruct::Type & holdTimeLimitsStruct)
+{
+    VerifyOrReturnError(kInvalidEndpointId != endpoint, CHIP_ERROR_INVALID_ARGUMENT);
+
+    size_t endpointIndex = 0;
+    ReturnErrorOnFailure(FindHoldTimeLimitsIndex(endpoint, endpointIndex));
+
+    mHoldTimeLimitsStructs[endpointIndex] = holdTimeLimitsStruct;
+
+    return CHIP_NO_ERROR;
+}
+
+/// @brief Returns the index of the HoldTimeLimits associated to an endpoint
+/// @param[in] endpoint target endpoint
+/// @param[out] endpointIndex index of the corresponding HoldTimeLimits for an endpoint
+/// @return CHIP_NO_ERROR or CHIP_ERROR_NOT_FOUND, CHIP_ERROR_INVALID_ARGUMENT if invalid endpoint
+CHIP_ERROR HoldTimeLimitsManager::HoldTimeLimits::FindHoldTimeLimitsIndex(EndpointId endpoint, size_t & endpointIndex)
+{
+    VerifyOrReturnError(kInvalidEndpointId != endpoint, CHIP_ERROR_INVALID_ARGUMENT);
+
+    uint16_t index = emberAfGetClusterServerEndpointIndex(endpoint, OccupancySensing::Id,
+                                                          MATTER_DM_OCCUPANCY_SENSING_CLUSTER_SERVER_ENDPOINT_COUNT);
+
+    if (index < ArraySize(mHoldTimeLimitsStructs))
+    {
+        endpointIndex = index;
+        return CHIP_NO_ERROR;
+    }
+    return CHIP_ERROR_NOT_FOUND;
+}
+
+HoldTimeLimitsManager HoldTimeLimitsManager::mInstance;
+
+HoldTimeLimitsManager & HoldTimeLimitsManager::Instance()
+{
+    return mInstance;
+}
+
+CHIP_ERROR HoldTimeLimitsManager::Init()
+{
+    // Prevents re-initializing
+    VerifyOrReturnError(!mIsInitialized, CHIP_ERROR_INCORRECT_STATE);
+
+    for (size_t i = 0; i <= kOccupancySensingServerMaxEndpointCount; i++)
+    {
+        if (emberAfContainsServer(EndpointId(i), OccupancySensing::Id))
+        {
+            Structs::HoldTimeLimitsStruct::Type holdTimeLimitsInit;
+
+            // Set up some sane initial values for hold time limits structures
+            holdTimeLimitsInit.holdTimeMin     = 1;
+            holdTimeLimitsInit.holdTimeMax     = 300;
+            holdTimeLimitsInit.holdTimeDefault = 10;
+
+            HoldTimeLimitsManager::Instance().SetHoldTimeLimitsStruct(EndpointId(i), holdTimeLimitsInit);
+        }
+    }
+
+    VerifyOrReturnError(registerAttributeAccessOverride(this), CHIP_ERROR_INCORRECT_STATE);
+
+    mIsInitialized = true;
+    return CHIP_NO_ERROR;
+}
+
+// AttributeAccessInterface
+CHIP_ERROR HoldTimeLimitsManager::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    switch (aPath.mAttributeId)
+    {
+    case Attributes::HoldTimeLimits::Id: {
+
+        Structs::HoldTimeLimitsStruct::Type * holdTimeLimitsStruct =
+            HoldTimeLimitsManager::Instance().mHoldTimeLimits.GetHoldTimeLimitsStruct(aPath.mEndpointId);
+        Structs::HoldTimeLimitsStruct::Type res;
```

Comment:
Why do we need an extra copy?  Why can't we just `aEncoder.Encode(*holdTimeLimitsStruct)`?

Separately, how confident are we that `holdTimeLimitsStruct` is never null here?

Reply:
Added check on null pointer, and got rid of the extra copy, please check